src/tools/rbd_mirror/group_replayer/Replayer.cc

```diff
@@ -346,26 +346,61 @@ void Replayer<I>::load_local_group_snapshots() {
     }
   }
 
-  //FIXME: This is not accessed under a lock
+
+  std::unique_lock locker{m_lock};
   if (!m_local_group_snaps.empty()) {
+    if (!m_prune_snap_ids.empty()) {
+      clean_unnecessary_mirror_group_snaps(locker);
+    }
     for (auto &local_snap : m_local_group_snaps) {
-      if (local_snap.state == cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE) {
-        continue;
-      }
       // do not forward until previous local mirror group snapshot is COMPLETE
-      validate_image_snaps_sync_complete(local_snap.id);
       auto snap_type = cls::rbd::get_group_snap_namespace_type(
           local_snap.snapshot_namespace);
       if (snap_type == cls::rbd::GROUP_SNAPSHOT_NAMESPACE_TYPE_MIRROR
           || m_retry_validate_snap) {
+        if (local_snap.state == cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE) {
+          continue;
+        }
         m_retry_validate_snap = false;
-        schedule_load_group_snapshots();
-        return;
+        validate_image_snaps_sync_complete(local_snap.id, locker);
+        clean_unnecessary_mirror_group_snaps(locker);
+      } else {
+        // its a user snapshot
+        bool found = false;
+        for (auto &remote_snap : m_remote_group_snaps) {
+          if (local_snap.id == remote_snap.id) {
+            found = true;
+            break;
+          }
+        }
+        if (found) { // user snap added
+          if (local_snap.state == cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE) {
+            continue;
+          }
+          validate_image_snaps_sync_complete(local_snap.id, locker);
+          clean_unnecessary_mirror_group_snaps(locker);
+          continue;
+        } else { // user snap removed
+          auto local_snap_id = local_snap.id;
```

Comment:
The snap removal might not be captured by the current mirror snapshot.  This is the same issue @idryomov brought up in the discussion.

Reply:
I mean that there may be multiple mirror snaps and the group snap change might be in the second mirror snap, not the first.